p2p/src/main/kotlin/maru/p2p/discovery/MaruDiscoveryService.kt

```diff
@@ -0,0 +1,183 @@
+
+    discoveryNetworkBuilder.listen(p2pConfig.ipAddress, p2pConfig.discoveryPort.toInt())
+    discoveryNetworkBuilder.secretKey(privateKey)
+    discoveryNetworkBuilder.localNodeRecord(localNodeRecord())
+    discoveryNetworkBuilder.bootnodes(bootnodes)
+
+    discoverySystem = discoveryNetworkBuilder.build()
+  }
+
+  fun start() {
+    discoverySystem
+      .start()
+      .thenRun {
+        this.bootnodeRefreshTask =
+          delayedExecutor.runWithFixedDelay(
+            { this.pingBootnodes(bootnodes) },
+            BOOTNODE_REFRESH_DELAY,
+            { error: Throwable ->
+              log.error(
+                "Failed to contact discovery bootnodes",
+                error,
+              )
+            },
+          )
+      }
+  }
+
+  fun stop() {
+    bootnodeRefreshTask.cancel()
+    discoverySystem.stop()
+  }
+
+  fun updateForkId(forkId: ForkId) { // TODO: Need to call this when the fork id changes
+    discoverySystem.updateCustomFieldValue(
+      FORK_ID_FIELD_NAME,
+      forkId.bytes,
+    )
+  }
+
+  fun searchForPeers(): SafeFuture<Collection<MaruDiscoveryPeer>> =
+    SafeFuture
+      .of(discoverySystem.searchForNewPeers())
+      // The current version of discovery doesn't return the found peers but next version will
+      .thenApply { getKnownPeers() }
+
+  fun getKnownPeers(): Collection<MaruDiscoveryPeer> =
+    discoverySystem
+      .streamLiveNodes()
+      .map { node: NodeRecord ->
+        convertNodeRecordToDiscoveryPeer(node)
+      }.filter { checkPeer(it) }
+      .toList()
+
+  fun getLocalNodeRecord(): NodeRecord = discoverySystem.localNodeRecord
+
+  private fun convertNodeRecordToDiscoveryPeer(node: NodeRecord): MaruDiscoveryPeer {
+    val forkIdBytes = node.get(FORK_ID_FIELD_NAME) as? Bytes?
+    return MaruDiscoveryPeer(
+      (node.get(EnrField.PKEY_SECP256K1) as? Bytes),
+      node.nodeId,
+      node.tcpAddress.orElse(null),
+      Optional.ofNullable(forkIdBytes),
+    )
+  }
+
+  private fun pingBootnodes(bootnodeRecords: List<NodeRecord>) {
+    bootnodeRecords.forEach(
+      Consumer { bootnode: NodeRecord? ->
+        SafeFuture
+          .of(discoverySystem.ping(bootnode))
+          .whenComplete { _, e ->
+            if (e != null) {
+              log.warn("Bootnode {} is unresponsive", bootnode)
+            }
+          }
+      },
+    )
+  }
+
+  private fun localNodeRecord(): NodeRecord {
+    val nodeRecordBuilder: NodeRecordBuilder =
+      NodeRecordBuilder()
+        .secretKey(privateKey)
+        .seq(UInt64.ONE)
+        .address(
+          p2pConfig.ipAddress,
+          p2pConfig.discoveryPort.toInt(),
+          p2pConfig.port.toInt(),
+        ).customField(FORK_ID_FIELD_NAME, forkIdProvider.invoke().bytes)
+    // TODO: do we want more custom fields to identify version/topics/role/something else?
+
+    return nodeRecordBuilder.build()
+  }
+
+  private fun checkPeer(peer: MaruDiscoveryPeer): Boolean {
+    if (peer.nodeIdBytes == null ||
```

Comment:
I don't see a point of this. Can't we avoid this check by making all the fields non-optional?

Reply:
The node record that is used to create the MaryDiscoveryPeer is created from data sent to us from the other peer.  They can send anything.

Reply:
Maybe instead of a map after `streamLiveNodes` we can use flatMap and reject all the items that don't make sense for us